### PR TITLE
perf(ops): Monomorphic sync op calls

### PR DESCRIPTION
**Pull request by aapoalas**
Welcome to better optimised op calls! Currently `opSync` is called with parameters of every type and count. This most definitely makes the call megamorphic. Additionally, it seems that spread params leads to V8 not being able to optimise the calls quite as well (apparently Fast Calls cannot be used with spread params).

Monomorphising op calls should lead to some improved performance. Now that unwrapping of sync ops results is done on Rust side, this is pretty simple:
```ts
opSync("op_foo", param1, param2);
// -> turns to
ops.op_foo(param1, param2);
```
This means sync op calls are now just directly calling the native binding function. When V8 Fast API Calls are enabled, this will enable those to be called on the optimised path.

Monomorphising async ops likely requires using callbacks and is left as an exercise to the reader.

## Performance

Comparison of `nop` op in a vacuum:
```
running 2 tests
test bench_op_nop        ... bench:      37,315 ns/iter (+/- 5,121)
test bench_op_nop_opSync ... bench:      37,925 ns/iter (+/- 4,071)
```

Comparison of FFI related ops:
```
benchmark                            time (avg)             (min … max)
-----------------------------------------------------------------------
ops.op_ffi_cstr_read(ptr)        121.16 ns/iter  (102.2 ns … 249.56 ns)
opSync('op_ffi_cstr_read', ptr)   131.3 ns/iter (111.42 ns … 302.48 ns)
ops.op_ffi_ptr_of(buf)           140.96 ns/iter (125.56 ns … 246.76 ns)
opSync('op_ffi_ptr_of', buf)     184.02 ns/iter (162.43 ns … 411.55 ns)
```

## Discussion

There are good things about this:
1. Performance
2. Picking the fruits created by the ops macro PR.

Downsides are:

1. None

### CODE REVIEW

File: cli/bench/deno_common.js [hunk 1]
```diff
@@ -6,7 +6,7 @@ Deno.bench("date_now", { n: 5e5 }, () => {
 });
 
 // Void ops measure op-overhead
-Deno.bench("op_void_sync", { n: 1e7 }, () => Deno.core.opSync("op_void_sync"));
+Deno.bench("op_void_sync", { n: 1e7 }, () => Deno.core.opsop_void_sync());
```
Comment (Phosra) [inline cli/bench/deno_common.js:9]:
```suggestion
Deno.bench("op_void_sync", { n: 1e7 }, () => Deno.core.ops.op_void_sync());
```

File: bench_util/benches/op_baseline.rs [hunk 2]
```diff
@@ -30,11 +30,11 @@ async fn op_pi_async() -> i64 {
 }
 
 fn bench_op_pi_json(b: &mut Bencher) {
-  bench_js_sync(b, r#"Deno.core.opSync("op_pi_json", null);"#, setup);
+  bench_js_sync(b, r#"Deno.core.ops.op_pi_json();"#, setup);
 }
 
 fn bench_op_nop(b: &mut Bencher) {
-  bench_js_sync(b, r#"Deno.core.opSync("op_nop", null, null, null);"#, setup);
```
Comment (Phosra) [inline bench_util/benches/op_baseline.rs:37]: Shouldn't the `null`s be preserved?
Was this measuring argument discarding speed? (lol)

File: core/01_core.js [hunk 3]
```diff
@@ -238,40 +239,44 @@
...
-      null,
-      "op_set_promise_reject_callback",
-    ),
+    print: (msg, isErr) => unwrapOpResult(ops.op_print(msg, isErr)),
+    setMacrotaskCallback: (fn) =>
+      unwrapOpResult(ops.op_set_macrotask_callback(fn)),
+    setNextTickCallback: (fn) =>
+      unwrapOpResult(ops.op_set_next_tick_callback(fn)),
+    runMicrotasks: () => ops.op_run_microtasks(),
+    hasTickScheduled: () => ops.op_has_tick_scheduled(),
```
Comment (Phosra) [inline core/01_core.js:253]:
```suggestion
    runMicrotasks: ops.op_run_microtasks,
    hasTickScheduled: ops.op_has_tick_scheduled,
```
?
(and for others?)

File: core/serialize_deserialize_test.js [hunk 4]
```diff
@@ -63,12 +61,11 @@ function main() {
   ];
 
   assertArrayEquals(
-    Deno.core.opSync("op_serialize", circularObject),
+    Deno.core.ops.op_serialize(circularObject),
```
Comment (Phosra) [inline core/serialize_deserialize_test.js:64]: `op_serialize` and `op_deserialize` are fallible.

**Comment by Phosra** [comment]
IMO, It's reasonable to just do a swift inlining pass that ignores eluding `unwrapOpResult`, if it makes it easier to get a first iteration done here.

File: runtime/js/10_permissions.js [hunk 5]
```diff
@@ -5,7 +5,7 @@
   const {
     Event,
     EventTarget,
-    Deno: { core },
+    Deno: { core: { ops } },
```
Comment (Phosra) [inline runtime/js/10_permissions.js:8]:
```suggestion
    Deno: { core, core: { ops } },
```
This is causing the failures, they're `ReferenceError`s for `core`.

File: bench_util/benches/op_baseline.rs (hunk 2 above)
Comment (aapoalas) [reply bench_util/benches/op_baseline.rs:37]: I think these are just leftovers from when op call arity was fixed.

File: core/01_core.js (hunk 3 above)
Comment (aapoalas) [reply core/01_core.js:253]: `ops` cannot actually be directly referenced. It seems that this JS run even when Deno isn't really being set up, eg. in tests or such I think.

In (some?) of those cases `ops` will be undefined and these sorts of lines will throw an error.

File: core/serialize_deserialize_test.js (hunk 4 above)
Comment (aapoalas) [reply core/serialize_deserialize_test.js:64]: These are testing array equality of valid results, ie. here the calls will not throw => there's no need for the test to use `unwrapOpResult`.

**Comment by aapoalas** [comment]
> If it makes it easier to get a first iteration done here, IMO it's reasonable to just do a swift inlining pass that ignores eluding `unwrapOpResult`.

Yeah, I think a lot of the files I'll just be doing that.

File: cli/bench/deno_common.js [hunk 6]
```diff
@@ -6,7 +6,11 @@ Deno.bench("date_now", { n: 5e5 }, () => {
 });
 
 // Void ops measure op-overhead
-Deno.bench("op_void_sync", { n: 1e7 }, () => Deno.core.ops.op_void_sync());
+Deno.bench(
+  "op_void_sync",
+  { n: 1e7 },
+  () => Deno.core.unwrapOpResult(Deno.core.ops.op_void_sync()),
```
Comment (Phosra) [inline cli/bench/deno_common.js:12]: `op_void_sync` is infallible. Is this to measure the conditional unwrap?

File: test_ffi/tests/bench.js [hunk 7]
```diff
@@ -1,7 +1,8 @@
 // Copyright 2018-2022 the Deno authors. All rights reserved. MIT license.
 // deno-lint-ignore-file
 
-const targetDir = Deno.execPath().replace(/[^\/\\]+$/, "");
+//const targetDir = Deno.execPath().replace(/[^\/\\]+$/, "");
+const targetDir = "/home/aapoalas/build/deno/target/release";
```
Comment (Phosra) [inline test_ffi/tests/bench.js:5]: Should this line be reverted?

**Review by Phosra — CHANGES_REQUESTED** [review]

File: cli/bench/deno_common.js (hunk 6 above)
Comment (aapoalas) [reply cli/bench/deno_common.js:12]: This is indeed meant to measure the unwrap together with the op call itself. Since only infallible ops that don't take parameters are truly infallible, it stands to reason that the unwrapping is an integral part of all ops and should be measured here.

File: test_ffi/tests/bench.js (hunk 7 above)
Comment (aapoalas) [reply test_ffi/tests/bench.js:5]: Yes.

**Review by Phosra — APPROVED** [review]

**Comment by aapoalas** [comment]
So, apparently this is ready for review and merge if nothing out of order is found. The tests failed once in parsing strace output but the failure did not repeat on another go around, so apparently it was a flake.

File: test_ffi/tests/test.js [hunk 8]
```diff
@@ -221,7 +217,11 @@ if (!(status & (1 << 4))) {
 dylib.symbols.print_buffer(ptr0, 8);
 const ptrView = new Deno.UnsafePointerView(ptr0);
 const into = new Uint8Array(6);
+// Make sure the buffer does not get collected
+globalThis.into = into;
```
Comment (bartlomieju) [inline test_ffi/tests/test.js:221]: This seems strange, can you elaborate what's the reason for that?

File: test_ffi/tests/thread_safe_test_worker.js [hunk 9]
```diff
@@ -35,7 +35,7 @@ self.addEventListener("message", ({ data }) => {
   } else if (data === "call") {
     dylib.symbols.call_stored_function();
   } else if (data === "unref") {
-    callback.unref();
+    callback.close();
```
Comment (bartlomieju) [inline test_ffi/tests/thread_safe_test_worker.js:38]: Seems like an unrelated change

File: test_ffi/tests/thread_safe_test.js [hunk 10]
```diff
@@ -90,7 +90,10 @@ cleanupCallback.ref();
 
 function cleanup() {
   cleanupCallback.unref();
-  console.log("Isolate should now exit");
+  // Keep isolate alive for a bit more to make sure everything finishes properly.
+  setTimeout(() => {
+    console.log("Isolate should now exit");
+  }, 100);
```
Comment (bartlomieju) [inline test_ffi/tests/thread_safe_test.js:96]: Ditto

File: test_ffi/tests/test.js (hunk 8 above)
Comment (aapoalas) [reply test_ffi/tests/test.js:221]: I think I was trying to get rid of the flake in FFI tests, thinking that it might've been about V8 garbage collecting a buffer while a pointer number from `Deno.UnsafePointer.of(into)` is still being used.

However, looking at this now it's not definitely necessary. I'll remove these changes.

File: test_ffi/tests/thread_safe_test_worker.js (hunk 9 above)
Comment (aapoalas) [reply test_ffi/tests/thread_safe_test_worker.js:38]: This was me trying to, again, get rid of the flake in FFI tests.

File: test_ffi/tests/thread_safe_test.js (hunk 10 above)
Comment (aapoalas) [reply test_ffi/tests/thread_safe_test.js:96]: Ditto :)

File: test_ffi/tests/test.js (hunk 8 above)
Comment (aapoalas) [reply test_ffi/tests/test.js:221]: I'll leave the `.pointer` changes as the `ptr()` function doesn't really serve much of a purpose.

**Review by littledivy — APPROVED** [review]
love it!

**Review by bartlomieju — APPROVED** [review]
LGTM, thank you @aapoalas, great work
